### PR TITLE
Pretty printing

### PR DESCRIPTION
**Pull request by blitz-1306**
## Dependence
- Requires #64 to be merged first.

## Changes
- [x] Introduced `PPAble`, `PPish` types and `pp()` function for pretty-printing primitive and complex values.
- [x] Tweaked `details` argument of `assert()` to accept `PPish` array.
- [x] Added related tests.

Regards.

### CODE REVIEW

File: src/misc/pretty_printing.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        typeof value === "boolean" ||
+        typeof value === "bigint"
+    ) {
+        return String(value);
+    }
+
+    if (isPPAble(value)) {
+        return value.pp();
+    }
+
```
Comment (cd1m0) [inline src/misc/pretty_printing.ts:35]: Could we add the following ifs here (to allow for pp() to handle (potentially nested) arrays, maps and iterables):

```
if (value instanceof Array) {
  return ppArr(value);
}

if (<value is an Iterable>) {
   return ppIter(value);
}

if (value instanceof Set) {
   return ppSet(value);
}

if (value instanceof Map) {
   return ppMap(value);
}
```

After we add this, we should add tests that we can pretty-print types like `string[][]`, `Map<string, Map<number, number>>`, `Set<string>[][]` etc...

**Review by cd1m0 — CHANGES_REQUESTED** [review]
Just 1 comment :)

File: src/misc/utils.ts [hunk 2]
```diff
@@ -29,13 +29,13 @@ export function assert(
             let part: string;
 
             if (detail instanceof ASTNode) {
-                part = detail.type + " #" + detail.id;
```
Comment (cd1m0) [inline src/misc/utils.ts:32]: This if case can be removed (its already handled in pp()).

File: src/misc/utils.ts [hunk 3]
```diff
@@ -29,13 +29,13 @@ export function assert(
...
+                part = pp(detail);
 
                 nodes.push(detail);
-            } else if (detail instanceof TypeNode) {
-                part = detail.pp();
+            } else if (detail instanceof Array) {
```
Comment (cd1m0) [inline src/misc/utils.ts:35]: This case can be removed after we add the additional ifs in pp().

**Review by cd1m0 — CHANGES_REQUESTED** [review]

File: src/misc/utils.ts (hunk 3 above)
Comment (blitz-1306) [reply src/misc/utils.ts]: Addressed in https://github.com/ConsenSys/solc-typed-ast/commit/6e612113efada422eb324ed77841541855af8c10

File: src/misc/utils.ts (hunk 2 above)
Comment (blitz-1306) [reply src/misc/utils.ts:32]: Addressed in https://github.com/ConsenSys/solc-typed-ast/commit/6e612113efada422eb324ed77841541855af8c10

File: src/misc/pretty_printing.ts (hunk 1 above)
Comment (blitz-1306) [reply src/misc/pretty_printing.ts:47]: Addressed in https://github.com/ConsenSys/solc-typed-ast/commit/6e612113efada422eb324ed77841541855af8c10.

Still, we should be careful with possible recursion here.

**Review by cd1m0 — APPROVED** [review]
LGTM

**Pull request merged**